[functorch] Remove internal assert in index_put batch rule

qqaatw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100516

Fixes #94630

Context: https://github.com/pytorch/pytorch/issues/94630#issuecomment-1518946929

I'm not sure how to correctly test it using pure `vmap` without joining `jacrev` here.

cc @zou3519

qqaatw committed:
[functorch] Remove internal assert in index_put batch rule

[ghstack-poisoned]

zou3519 commented:
Here's a minimal repro without jacrev:
```
import torch

def backward(x, gy):
    view = x.view(-1, 3, 3)
    clone = view.clone()
    lt = clone < 1e-09
    zeros = torch.zeros([])
    index_put = torch.ops.aten.index_put.default(gy, [lt], zeros)
    return index_put

B = 2
x = torch.randn(3, 3)
gy = torch.randn(B, 1, 3, 3)

torch.vmap(backward, (None, 0))(x, gy)
```

zou3519 commented on aten/src/ATen/functorch/BatchRulesScatterOps.cpp:
```diff
@@ -568,7 +568,6 @@ static Tensor maybe_permute_values(
   bool advanced_indices_are_adjacent = are_advanced_indices_adjacent(orig_indices);
   auto num_leading_nones = get_num_leading_nones(orig_indices);
   auto max_index_dim = get_max_index_logical_dim(orig_indices, orig_indices_bdims);
-  TORCH_INTERNAL_ASSERT(values.dim() >= num_leading_nones + max_index_dim);
```
So, my understanding of this assert is that values.dim() should actually be equal to `num_leading_nones + max_index_dim` and this is an invariant that holds after `index_put_batch_rule_helper`.

It looks like `index_put_batch_rule_helper` doesn't actually handle the case where values_bdim is None? https://github.com/pytorch/pytorch/blob/3e18d3958be3dfcc36d3ef3c481f064f98ebeaf6/aten/src/ATen/functorch/BatchRulesScatterOps.cpp#L428-L454 (cc @kshitij12345 since you wrote this and may remember it more than I do, does that sound right?)

If that's correct then the fix would be instead be to fix `index_put_batch_rule_helper`.

kshitij12345 replied:
I don't remember exact details 😓 but will spend sometime reading it again tomorrow. Thanks!

kshitij12345 replied:
At a quick glance, I agree that problem seems to be with `index_put_batch_rule_helper` returning incorrectly shaped values.

kshitij12345 replied:
Actually, I think assert is not correct.

```python

import torch
import torch.utils._pytree as pytree

torch.manual_seed(133)

def loop(op, in_dims, out_dim, batch_size, *batched_args, **kwarg_values):
    outs = []
    out_spec = None
    for idx in range(batch_size):
        flat_args, args_spec = pytree.tree_flatten(batched_args)
        flat_dims, dims_spec = pytree.tree_flatten(in_dims)
        assert(args_spec == dims_spec)
        new_args = [a.select(in_dim, idx) if in_dim is not None else a for a, in_dim in zip(flat_args, flat_dims)]
        out = op(*pytree.tree_unflatten(new_args, args_spec), **kwarg_values)
        flat_out, out_spec = pytree.tree_flatten(out)
        outs.append(flat_out)

    outs = zip(*outs)
    result = [torch.stack(out_lst) for out_lst in outs]
    return pytree.tree_unflatten(result, out_spec)

def backward(x, gy):
    lt = x < 1e-09
    zeros = torch.zeros([])
    indexed_gy = torch.ops.aten.index(gy, [lt])
    print(indexed_gy.shape, "INDEX SHAPE")  # [4] or [B, 4]
    index_put = torch.ops.aten.index_put.default(gy, [lt], zeros, True)
    return index_put

B = 2
x = torch.randn(1, 3, 3)
gy = torch.randn(B, 1, 3, 3)

o = loop(backward, (None, 0), 0, B, x, gy)
print("*****"*8)
a = torch.func.vmap(backward, (None, 0))(x, gy)

torch.testing.assert_close(o, a)
```
From the above example, I think the shape of value should be broadcastable to the indexed shape and not necessarily the leading none + max logical index shape.

zou3519 replied:
Thanks for the example. I agree the assert is bogus, let's remove it

zou3519 commented:
@qqaatw could you add the test case before shipping? Ideally we would add it to https://github.com/pytorch/pytorch/blob/7a15e823885ddad1696e507a642fc89029a75009/test/functorch/functorch_additional_op_db.py#L280 , though I'm a bit concerned because those tests look comprehensive but didn't catch this bug.

qqaatw committed:
Update on "[functorch] Remove internal assert in index_put batch rule"


Fixes #94630

Context: https://github.com/pytorch/pytorch/issues/94630#issuecomment-1518946929

I'm not sure how to correctly test it using pure `vmap` without joining `jacrev` here.

cc zou3519 

[ghstack-poisoned]

qqaatw committed:
Update on "[functorch] Remove internal assert in index_put batch rule"


Fixes #94630

Context: https://github.com/pytorch/pytorch/issues/94630#issuecomment-1518946929

I'm not sure how to correctly test it using pure `vmap` without joining `jacrev` here.

cc zou3519 

[ghstack-poisoned]

qqaatw commented:
@zou3519 I think we can't specify boolean mask index outside the function to be v-mapped, i.e. `x < 1e-09` should be placed inside the function; otherwise, the index will be specified a batch dim to test, which will hit the dynamic shape check. (See the CI failures)

zou3519 commented:
> So maybe not adding a test sample in the op_db but adding a manual test in test_vmap.py instead?

Manual test sgtm

qqaatw committed:
Update on "[functorch] Remove internal assert in index_put batch rule"


Fixes #94630

Context: https://github.com/pytorch/pytorch/issues/94630#issuecomment-1518946929

I'm not sure how to correctly test it using pure `vmap` without joining `jacrev` here.

cc zou3519 

[ghstack-poisoned]

qqaatw commented:
> > So maybe not adding a test sample in the op_db but adding a manual test in test_vmap.py instead?
> 
> Manual test sgtm

Done.

zou3519 reviewed (approved):
thank you!
